src/Google.Api.Gax/PageStreamingAsync.cs

```diff
@@ -0,0 +1,249 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+// Interfaces and implementations of asynchronous page streaming. These are gathered into a single file
+// for convenience.
+
+namespace Google.Api.Gax
+{
+    /// <summary>
+    /// An asynchronous sequence of resources obtained via API responses. Application code
+    /// can treat this as a simple sequence (with API calls automatically being made
+    /// lazily as more results are required), or call <see cref="AsPages"/> to retrieve
+    /// a page at a time, potentially with additional information.
+    /// </summary>
+    public interface IPagedAsyncEnumerable<TResponse, TResource> : IAsyncEnumerable<TResource>
+    {
+        /// <summary>
+        /// Returns the sequence of API responses, each of which contributes a page of
+        /// resources to this sequence.
+        /// </summary>
+        /// <returns>An asynchronous sequence of API responses, each containing a page of resources.</returns>
+        IResponseAsyncEnumerable<TResponse, TResource> AsPages();
+    }
+
+    /// <summary>
+    /// An asynchronous sequence of API responses, each of which contains a page of resources and
+    /// potentially additional information (depending on the API).
+    /// </summary>
+    /// <typeparam name="TResponse">The API response type. Each response contains a page of resources.</typeparam>
+    /// <typeparam name="TResource">The resource type contained within the response.</typeparam>
+    public interface IResponseAsyncEnumerable<TResponse, TResource> : IAsyncEnumerable<TResponse>
+    {
+        /// <summary>
+        /// Begins lazily iterating over the sequence of pages, with an iterator allowing the page size to
+        /// be specified on each step.
+        /// </summary>
+        /// <returns>An asynchronous iterator over the pages in the sequence.</returns>
+        new IResponseAsyncEnumerator<TResponse> GetEnumerator();
+
+        /// <summary>
+        /// Creates a lazily-evaluated asynchronous sequence of pages of resources, where each page other
+        /// than the final one is guaranteed to contain exactly <paramref name="pageSize"/> resources.
+        /// </summary>
+        /// <remarks>
+        /// "Natural" pages returned by the API may contain a smaller number of resources than requested.
+        /// For example, a request for a page with 100 resources may return a page with 80 resources but
+        /// a next page token for more to be retrieved. This is suitable for batch-processing, but not
+        /// for user-visible paging such as in a web application, where fixed-size pages are expected.
+        /// </remarks>
+        /// <param name="pageSize">The page size. Must be greater than 0.</param>
+        /// <returns>A lazily-evaluated asynchronous sequence of fixed-size pages.</returns>
+        IAsyncEnumerable<FixedSizePage<TResource>> WithFixedSize(int pageSize);
```

Comment:
Just wondering of this would be better call "OfFixedSize"?
ListTopics().AsPages().OfFixedSize(n)